tests/api/test_type_guard.py

```diff
@@ -0,0 +1,210 @@
+# SPDX-FileCopyrightText: 2023 Mewbot Developers <mewbot@quicksilver.london>
+#
+# SPDX-License-Identifier: BSD-2-Clause
+
+"""
+Tests check_input_event_against_sig_types using a dummy Trigger.
+"""
+
+from typing import Any, Union
+
+from mewbot.api.v1 import Trigger, InputEvent
+from mewbot.api.type_guard import check_input_event_against_sig_types
+
+from mewbot.io.http import IncomingWebhookEvent
+from mewbot.io.socket import SocketInputEvent
+
+
+class ATestTrigger(Trigger):
+    """
+    Dummy trigger for testing.
+    """
+
+    @staticmethod
+    def consumes_inputs() -> set[type[InputEvent]]:
+        """
+        Consumes two radically different events.
+
+        :return:
+        """
+        return {SocketInputEvent, IncomingWebhookEvent}
+
+    @check_input_event_against_sig_types
+    def matches(self, event: Union[SocketInputEvent, IncomingWebhookEvent]) -> bool:
```

Comment:
We should have at least one test using `X | Y` syntax because it is slightly different. `from __future__ import annotations` should allow it in python 3.9, but I might recommend having a separate test file for it -- that we can have better confidence that both forms work as expected.

Reply:
`from __future__ import annotations` does not seem to allow it in 3.9 - I'm gonna gate it behind a version check.